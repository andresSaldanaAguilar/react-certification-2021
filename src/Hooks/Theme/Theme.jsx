import React, { createContext, useContext, useState } from 'react';

const ThemeStateContext = createContext();
const ThemeDispatchContext = createContext();

function ThemeProvider({ children }) {
  const [theme, setTheme] = useState('light');
  return (
    <ThemeStateContext.Provider value={theme}>
      <ThemeDispatchContext.Provider value={setTheme}>
        {children}
      </ThemeDispatchContext.Provider>
    </ThemeStateContext.Provider>
  );
}
function useThemeState() {
  const context = useContext(ThemeStateContext);
  if (context === undefined) {
    throw new Error('useThemeState must be used within a ThemeProvider');
  }
  return context;
}
function useThemeDispatch() {
  const context = useContext(ThemeDispatchContext);
  if (context === undefined) {
    throw new Error('useThemeDispatch must be used within a ThemeProvider');
  }
  return context;
}
export { ThemeProvider, useThemeState, useThemeDispatch };
