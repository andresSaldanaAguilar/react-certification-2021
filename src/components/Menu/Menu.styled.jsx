import { AppBar, InputBase, Toolbar } from '@material-ui/core';
import styled from 'styled-components';
import { fade } from '@material-ui/core/styles';

export const CustomAppBar = styled(AppBar)`
  background-color: ${(props) => props.theme.colors.blueGrey} !important;
`;

export const CustomToolbar = styled(Toolbar)`
  display: flex;
  justify-content: space-between;
`;

export const ToolbarSection = styled.div`
  display: flex;
`;

export const SearchContainer = styled.div`
  position: relative;
  border-radius: 5px;
  background-color: ${(props) => fade(props.theme.colors.white, 0.15)};
  ${ToolbarSection}:hover & {
    background-color: ${(props) => fade(props.theme.colors.white, 0.25)};
  }
  margin-right: 1em;
  width: 100%;
`;

export const SearchIconContainer = styled.div`
  padding: 0.5em;
  height: 100%;
  position: absolute;
  pointer-events: none;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const CustomInputBase = styled(InputBase)`
  margin: 0.25em;
  padding-left: 2em;
  width: 100%;
`;
