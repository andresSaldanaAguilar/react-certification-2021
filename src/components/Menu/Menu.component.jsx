import React, { useState } from 'react';
import { Menu, MenuItem, Typography } from '@material-ui/core';
import YouTubeIcon from '@material-ui/icons/YouTube';
import SearchIcon from '@material-ui/icons/Search';
import { useHistory, useLocation } from 'react-router-dom';
import {
  CustomAppBar,
  CustomButton,
  CustomIconButton,
  CustomInputBase,
  CustomToolbar,
  SearchContainer,
  SearchIconContainer,
  ThemeIcon,
  ToolbarSection,
} from './Menu.styled';
import { useSwitchTheme } from '../../Hooks/Theme/Theme';
import { useSession } from '../../Hooks/Session/Session';
import { useSearchDispatch } from '../../Hooks/Search/Search';

function HomeButton(history) {
  return (
    <CustomIconButton
      onClick={() => history.push('/')}
      edge="start"
      data-testid="HomeButton"
    >
      <YouTubeIcon />
      <Typography variant="body1" noWrap>
        MyTube
      </Typography>
    </CustomIconButton>
  );
}

function SearchBar(path) {
  const setSearch = useSearchDispatch();
  const homePage = '/';
  return (
    path === homePage && (
      <SearchContainer data-testid="SearchBar">
        <SearchIconContainer>
          <SearchIcon />
        </SearchIconContainer>
        <CustomInputBase
          data-testid="SearchInput"
          placeholder="Search…"
          onChange={(e) => {
            setSearch(e.target.value);
          }}
        />
      </SearchContainer>
    )
  );
}

function handleLoginButton(userSession, history) {
  const { session, dispatchSession } = userSession;
  if (session.user) {
    dispatchSession({ type: 'logout' });
    history.push('/');
  } else {
    history.push('/login');
  }
}

function LoginSection(userSession, history) {
  const { session } = userSession;
  const [anchorEl, setAnchorEl] = useState(null);

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };

  return (
    <>
      <CustomButton onClick={handleClick}>
        <Typography variant="body1" noWrap>
          {session.user ? session.user.name : 'Not signed'} ▼
        </Typography>
      </CustomButton>
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={() => setAnchorEl(null)}
      >
        {session.user && (
          <MenuItem onClick={() => history.push('/starred')}>Starred Videos</MenuItem>
        )}
        <MenuItem onClick={() => handleLoginButton(userSession, history)}>
          {session.user ? 'Logout' : 'Login'}
        </MenuItem>
      </Menu>
    </>
  );
}

function NavBar() {
  const switchTheme = useSwitchTheme();
  const location = useLocation();
  const history = useHistory();
  const userSession = useSession();
  return (
    <>
      <CustomAppBar position="fixed" data-testid="CustomAppBar">
        <CustomToolbar>
          <ToolbarSection>
            {HomeButton(history)}
            {SearchBar(location.pathname)}
          </ToolbarSection>
          <ToolbarSection>
            <ThemeIcon fontSize="default" onClick={switchTheme} />
            {LoginSection(userSession, history)}
          </ToolbarSection>
        </CustomToolbar>
      </CustomAppBar>
    </>
  );
}

export default NavBar;
