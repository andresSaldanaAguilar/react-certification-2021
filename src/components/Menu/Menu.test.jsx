import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Menu from './Menu.component';
import Theme from '../App/App.styled';

describe('Menu Component Tests', () => {
  const setup = () => {
    const utils = render(
      <BrowserRouter>
        <Theme>
          <Menu />
        </Theme>
      </BrowserRouter>
    );
    return {
      ...utils,
    };
  };

  it('Should render the menu component', () => {
    const { getByTestId } = setup();
    expect(getByTestId('CustomAppBar')).toBeInTheDocument();
  });

  it('Should redirect to home', async () => {
    const { getByTestId } = setup();
    const menuButton = getByTestId('MenuButton');
    menuButton.click();
    /* TODO */
  });

  it('Should change the switch state', async () => {
    const { getByTestId } = setup();
    const themeSwitch = getByTestId('ThemeSwitch').getElementsByTagName('input')[0];
    themeSwitch.click();
    expect(themeSwitch.checked).toEqual(true);
    themeSwitch.click();
    expect(themeSwitch.checked).toEqual(false);
  });
});
