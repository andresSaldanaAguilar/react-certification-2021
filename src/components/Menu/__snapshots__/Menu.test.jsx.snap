// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component Tests Should render the menu component 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary sc-bdfBwQ fofWC mui-fixed MuiPaper-elevation4"
    data-testid="CustomAppBar"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular sc-gsTCUz dmzGAY MuiToolbar-gutters"
    >
      <div
        class="sc-dlfnbm CZBcd"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root title MuiIconButton-colorInherit MuiIconButton-edgeStart"
          data-testid="MenuButton"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 15l5.19-3L10 9v6m11.56-7.83c.13.47.22 1.1.28 1.9.07.8.1 1.49.1 2.09L22 12c0 2.19-.16 3.8-.44 4.83-.25.9-.83 1.48-1.73 1.73-.47.13-1.33.22-2.65.28-1.3.07-2.49.1-3.59.1L12 19c-4.19 0-6.8-.16-7.83-.44-.9-.25-1.48-.83-1.73-1.73-.13-.47-.22-1.1-.28-1.9-.07-.8-.1-1.49-.1-2.09L2 12c0-2.19.16-3.8.44-4.83.25-.9.83-1.48 1.73-1.73.47-.13 1.33-.22 2.65-.28 1.3-.07 2.49-.1 3.59-.1L12 5c4.19 0 6.8.16 7.83.44.9.25 1.48.83 1.73 1.73z"
              />
            </svg>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap"
            >
              MyTube
            </p>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="sc-hKgILt heiTUU"
        >
          <div
            class="sc-eCssSg clVwIQ"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <div
            class="MuiInputBase-root sc-jSgupP bKMman"
          >
            <input
              class="MuiInputBase-input"
              placeholder="Searchâ€¦"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="sc-dlfnbm CZBcd"
      >
        <span
          class="MuiSwitch-root sc-gKsewC"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase"
            data-testid="ThemeSwitch"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-4 MuiSwitch-input"
                type="checkbox"
                value=""
              />
              <span
                class="MuiSwitch-thumb"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track"
          />
        </span>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Login
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
